src/org/javaday/java8/misused/stream/NestedForEach.java

```diff
@@ -0,0 +1,56 @@
+package org.javaday.java8.misused.stream;
+
+import org.javaday.java8.misused.Annotations.Good;
+import org.javaday.java8.misused.Annotations.Ugly;
+
+import java.util.ArrayList;
+import java.util.HashSet;
+import java.util.List;
+import java.util.Set;
+
+import static java.util.stream.Collectors.toSet;
+
+//Taken from real code
+public class NestedForEach {
+
+    @Ugly
+    class NestedForEachWithExternalCollection {
+
+        public Set<String> retrievePromoRuleNames(List<BusinessTransaction> transactions) {
+            Set<String> ruleNamesWithPromo = new HashSet<>();
+            transactions.forEach(trnx -> trnx.getRules().stream()
+                    .filter(BusinessRule::isPromotion)
+                    .forEach(rule -> ruleNamesWithPromo.add(rule.getRuleName())));
+            return ruleNamesWithPromo;
+        }
+    }
+
+    @Good
+    class StreamOperationsChain {
+
+        public Set<String> retrievePromoRuleNames(List<BusinessTransaction> transactions) {
+            return transactions.stream()
+                    .flatMap(t -> t.getRules().stream())
+                    .filter(BusinessRule::isPromotion)
+                    .map(BusinessRule::getRuleName)
+                    .collect(toSet());
+        }
+    }
+
+    class BusinessTransaction {
```

Comment:
Ideally yes, but there could be several exceptions. As I told in previous comment, I'll leave it to you to merge NestedForEach and MultiLevelForEach))